rclrs/src/node/graph.rs

```diff
@@ -0,0 +1,568 @@
+use std::collections::HashMap;
+use std::ffi::{CStr, CString};
+use std::slice;
+
+use crate::rcl_bindings::*;
+use crate::{Node, RclrsError, ToResult};
+
+impl Drop for rmw_names_and_types_t {
+    fn drop(&mut self) {
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rcl_names_and_types_fini(self).ok().unwrap();
+        }
+    }
+}
+
+impl Drop for rmw_topic_endpoint_info_array_t {
+    fn drop(&mut self) {
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rmw_topic_endpoint_info_array_fini(self, &mut rcutils_get_default_allocator())
+                .ok()
+                .unwrap();
+        }
+    }
+}
+
+impl Drop for rcutils_string_array_t {
+    fn drop(&mut self) {
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rcutils_string_array_fini(self);
+        }
+    }
+}
+
+/// Topic names and types
+pub type TopicNamesAndTypes = HashMap<String, Vec<String>>;
+
+/// Topic endpoint information
+pub struct TopicEndpointInfo {
+    /// The name of the endpoint node
+    pub node_name: String,
+    /// The namespace of the endpoint node
+    pub node_namespace: String,
+    /// The type of the topic
+    pub topic_type: String,
+    /// the RMW endpoint type of the topic
+    pub endpoint_type: rmw_endpoint_type_t,
+}
+
+impl Node {
+    /// Get a list of topic names and types for publishers associated with a node
+    pub fn get_publisher_names_and_types_by_node(
+        &self,
+        node: &str,
+        namespace: &str,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        self.get_names_and_types_by_node_with_demangle_arg(
+            node,
+            namespace,
+            rcl_get_publisher_names_and_types_by_node,
+        )
+    }
+
+    /// Get a list of topic names and types for subscribers associated with a node
+    pub fn get_subscriber_names_and_types_by_node(
```

Comment:
Weird that this is called `subscribers` in rcl and not `subscriptions`. Could you check in `rclcpp` which one they use?

Reply:
https://github.com/ros2/rcl/blob/287ccd9ed06ff5bdded4dfb1130920d592a71bb7/rcl/include/rcl/graph.h#L142 calls it "subscribers", 
rclcpp does not seem to provide a wrapper for this function. I guess that raises the question - should I focus on "wrapping" the rcl functions, or imitating the rclcpp implementation?

Reply:
Yeah, I saw that it's called `subscribers` in that file, but that seems like a mistake to me, and I though maybe it seemed that way to the rclcpp authors too. But if they don't have that function, I'd say we can just call it `get_subscription_names_and_types_by_node()`.

We can provide functionality that rclcpp doesn't provide, so I'd say the focus is to cover the `rcl` API.